### PR TITLE
Repository overhaul changes

### PR DESCRIPTION
**Pull request by rexagod**
- Version bump to go1.17.
- Various configuration, grammar, and doc fixes.
- Various extraneous comments and unused components (and packages) removed.
- Operator SDK auto-gen configs upgraded based on kubebuilder/v3.
- Restructed codebase to maintain cohesion between similar logically-grouped code and isolation among different ones.

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

### CODE REVIEW

**Review by naveenpaul1 — CHANGES_REQUESTED** [review]
do not Merge it, these changes are not based on the latest master branch

**Comment by rexagod** [comment]
> these changes are not based on the latest master branch

How so?

**Comment by naveenpaul1** [comment]
@rexagod your changes are based on the latest code in the master branch, sorry for the confusion, you might need to hold these changes because I have done lots of changes in the last few changes in my local branch, and it will give lots of conflicts
There are a few changes which need more discussion

**Review by naveenpaul1 — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (rexagod) [reply Dockerfile:23]: For what?

File: bundle/manifests/mcg-osd-deployer.clusterserviceversion.yaml
Comment (rexagod) [reply bundle/manifests/mcg-osd-deployer.clusterserviceversion.yaml:250]: Yes.

File: bundle/manifests/mcg.openshift.io_managedmcgs.yaml
Comment (rexagod) [reply bundle/manifests/mcg.openshift.io_managedmcgs.yaml:47]: We don't have one right now, let's only use what's required.

File: Makefile
Comment (rexagod) [reply Makefile:36]: One comment is enough for one change. Please don't spam the PR.

File: controllers/managedmcg_controller.go
Comment (rexagod) [reply controllers/managedmcg_controller.go:157]: We'll rename when more conditions come in.

File: controllers/managedmcg_controller.go
Comment (rexagod) [reply controllers/managedmcg_controller.go:191]: I swapped it for the oneliner that was being used.

File: go.mod
Comment (rexagod) [reply go.mod:23]: This was added by golang itself due to the version bump to 1.17. I suggest we don't touch them.

File: LICENSE
Comment (rexagod) [reply LICENSE]: Added.

File: Dockerfile
Comment (rexagod) [reply Dockerfile:16]: Were they being used anywhere else (other than `controller`)?

File: config/crd/bases/mcg.openshift.io_managedmcgs.yaml
Comment (rexagod) [reply config/crd/bases/mcg.openshift.io_managedmcgs.yaml:49]: Where are we using it (right now)?

File: Dockerfile
Comment (nb-ohad) [reply Dockerfile:23]: @rexagod  We created the readiness prob in the first place so we could communicate the correct status to OCM to facilitate service status on the OCM UI (and let OCM know that the addon is full funcitoning)

I do belive the same reasoning also apply for the Data Federation addon

File: Dockerfile
Comment (naveenpaul1) [reply Dockerfile:16]: controller and util folders are completely different entities and I dont see any reason to put util inside controller folder

File: controllers/managedmcg_controller.go
Comment (rexagod) [reply controllers/managedmcg_controller.go:94]: Renamed to `initializeReconciler`.

File: controllers/managedmcg_controller.go
Comment (rexagod) [reply controllers/managedmcg_controller.go:204]: Moved.

File: bundle/manifests/mcg.openshift.io_managedmcgs.yaml
Comment (nb-ohad) [reply bundle/manifests/mcg.openshift.io_managedmcgs.yaml:47]: @rexagod If you are planning to push your metrics to the in-cluster monitoring stack (openshift-monitoring) you will probably get nacked by SREP, as in ROSA and OSD that monitoring stack is reserved for platform use only

File: bundle/manifests/mcg-osd-deployer.clusterserviceversion.yaml
Comment (rexagod) [reply bundle/manifests/mcg-osd-deployer.clusterserviceversion.yaml:20]: Reverted.

File: bundle/manifests/mcg-osd-deployer.clusterserviceversion.yaml
Comment (rexagod) [reply bundle/manifests/mcg-osd-deployer.clusterserviceversion.yaml:17]: I've reverted it because in my case the version I'm using is based on the latest commit, which is why it says unknown.

image

File: Dockerfile
Comment (naveenpaul1) [reply Dockerfile:23]: without readiness the add on in a ready state in the installed operator before noobaa is up and running, you will get green tick even if noobaa is not up also, this will create confusion, this is why we need readiness like the name suggest

File: Makefile
Comment (naveenpaul1) [reply Makefile:36]: to make it easy to revert

File: go.mod
Comment (naveenpaul1) [reply go.mod:23]: ok

**Comment by naveenpaul1** [comment]
// @v-harihar

File: Dockerfile
Comment (rexagod) [reply Dockerfile:16]: They are completely different entities because they were inside separate packages for no reason, since they are only being used in one part of the codebase. If this was not the case, we could have it inside `pkg` (as the go standard layout specifies: https://github.com/golang-standards/project-layout#pkg), or `internal` if it was repo-wide and not for use by external packages.

Comment (rexagod) [reply Dockerfile:16]: However this is not the case, and logically, at this point in time, those packages are moved to where they are needed.

File: Dockerfile
Comment (rexagod) [reply Dockerfile:23]: @nb-ohad Can you PLMK what part of OCM UI uses the readiness probe check? Is it the green tick? Or something else entirely? I could have missed it if it was the latter case since I only checked this on OCP.

@naveenpaul1 If there's a green tick even without readiness check, then what difference does the readiness probe, serving readyz 200 status make (I don't see any failed probe checks in the logs anyway)?

File: controllers/managedmcg_controller.go
Comment (naveenpaul1) [reply controllers/managedmcg_controller.go:159]: previous method name updateComponentStatus().
we are updating status.component{}

Comment (rexagod) [reply controllers/managedmcg_controller.go]: I've renamed to `updateNoobaaComponentStatus` to make it explicit that we are updating ManagedMCG's Noobaa component's status, not just any component.

File: Dockerfile
Comment (naveenpaul1) [reply Dockerfile:23]: @rexagod think it in this way, without the readiness prob you will get success mark even if the operator fails to start the noobaa, this is something we don't need, please let me know if you don't get this

File: Dockerfile
Comment (naveenpaul1) [reply Dockerfile:16]: @rexagod  I don't think it is good practice to have packages inside packages if it must and the codebase is huge, 
here we have not more than 4 folders and I don't see any problem putting under in the main folder,

Comment (rexagod) [reply Dockerfile:16]: If we do this now we'll have to refactor later. I'm not doing any of this from my own, I'm just going by what's defined in the standard.

File: Dockerfile
Comment (naveenpaul1) [reply Dockerfile:23]: I didnt get it

File: Dockerfile
Comment (rexagod) [reply Dockerfile:16]: Resolved over call: move templates to project root.

File: Dockerfile
Comment (rexagod) [reply Dockerfile:23]: Resolved over call: Use readiness server (as it was before).

**Comment by rexagod** [comment]
@aspandey While you're technically correct in suggesting this change, note that idiomatic Go enforces an error-first approach. Now, let's say we have something like this right now,
```go
_, err := fn1()
if err != nil {
	return err
}
_, err = fn2()
if err != nil {
	return err
}
return nil
```
and on the other hand, we have (the suggested change),
```
_, err := fn1()
if err != nil {
	return err
}
_, err = fn2()
return err
```

If you try to remove one function call and error block from the first example, you won't encounter any merge conflicts while rebasing on top of a PR that adds some other function there, BUT in the second case, the return param is a part of the second function call and if this goes on top of a PR that replaces this definition with its own function call and error block, there is going to be atleast one merge conflict.

This is the logic that convinced me to keep the fn calls and their error blocks isolated from the rest of the function body. But in case you are not convinced, or think this might be counter-intuitive, PLMK and I'll make those changes.

**Comment by aspandey** [comment]
> 
If it is a Go lang constraint or standard then I would not argue andd will be okay with the current code.
 
However, if avoiding merge conflict is the major reason behind following this coding style then I don't buy 
this argument. 
Whenever a developer makes changes in code, he is responsible for correctness of the code after removing/adding any part of code. If it requires changing little bit different code which can cause merge conflict then its okay. At the end the code should be 
correct and efficient.
We may be putting an extra effort once (to resolve merge conflict) while saving extra time and performance every time it gets executed.

File: controllers/managedmcg_controller.go
Comment (rexagod) [reply controllers/managedmcg_controller.go]: Fixed.

**Review by aspandey — APPROVED** [review]
LGMT, Reviewed only controller file and mainly code structure.

**Review by naveenpaul1 — APPROVED** [review]
LGTM

**Pull request merged**
